diff --git a/src/Xmobar/Plugins/Monitors/Disk.hs b/src/Xmobar/Plugins/Monitors/Disk.hs
index 0305d78..9ef652d 100644
--- a/src/Xmobar/Plugins/Monitors/Disk.hs
+++ b/src/Xmobar/Plugins/Monitors/Disk.hs
@@ -23,8 +23,7 @@ import Control.Exception (SomeException, handle)
 import Control.Monad (zipWithM)
 import qualified Data.ByteString.Lazy.Char8 as B
 import Data.List (isPrefixOf, find)
-import Data.Maybe (catMaybes)
-import System.Directory (canonicalizePath, doesFileExist)
+import System.Directory (canonicalizePath)
 import System.Console.GetOpt
 
 data DiskIOOpts = DiskIOOpts
@@ -98,15 +97,10 @@ type DevDataRef = IORef [(DevName, [Float])]
 mountedDevices :: [String] -> IO [(DevName, Path)]
 mountedDevices req = do
   s <- B.readFile "/etc/mtab"
-  parse `fmap` mapM mbcanon (devs s)
+  return $ parse $ devs s
   where
-    mbcanon (d, p) = doesFileExist d >>= \e ->
-                     if e
-                        then Just `fmap` canon (d,p)
-                        else return Nothing
-    canon (d, p) = do {d' <- canonicalizePath d; return (d', p)}
     devs = filter isDev . map (firstTwo . B.words) . B.lines
-    parse = map undev . filter isReq . catMaybes
+    parse = map undev . filter isReq
     firstTwo (a:b:_) = (B.unpack a, B.unpack b)
     firstTwo _ = ("", "")
     isDev (d, _) = "/dev/" `isPrefixOf` d
