diff --git a/src/mbyte.c b/src/mbyte.c
index 325527725..d93eedd87 100644
--- a/src/mbyte.c
+++ b/src/mbyte.c
@@ -1582,6 +1582,20 @@ utf_char2cells(int c)
 #endif
     };
 
+#if 1
+    if (c < 0x80)
+	return 1;
+
+    int n = wcwidth(c);
+    if (n <= 0)
+	return c > 0xFF ? 6 : 4;
+
+    int m = cw_value(c);
+    if (m != 0)
+	return m;
+
+    return n;
+#else
 #ifdef FEAT_EVAL
     // Use the value from setcellwidths() at 0x80 and higher, unless the
     // character is not printable.
@@ -1611,6 +1625,7 @@ utf_char2cells(int c)
 	return 2;
 
     return 1;
+#endif
 }
 
 /*
@@ -2685,6 +2700,9 @@ utf_iscomposing(int c)
     int
 utf_printable(int c)
 {
+#if 1
+    return wcwidth(c) > 0;
+#else
     // Sorted list of non-overlapping intervals.
     // 0xd800-0xdfff is reserved for UTF-16, actually illegal.
     static struct interval nonprint[] =
@@ -2695,6 +2713,7 @@ utf_printable(int c)
     };
 
     return !intable(nonprint, sizeof(nonprint), c);
+#endif
 }
 
 // Sorted list of non-overlapping intervals of all Emoji characters,
