
diff --git a/dlls/winex11.drv/systray.c b/dlls/winex11.drv/systray.c
index dde2feb..318e236 100644
--- a/dlls/winex11.drv/systray.c
+++ b/dlls/winex11.drv/systray.c
@@ -32,28 +32,6 @@ WINE_DEFAULT_DEBUG_CHANNEL(systray);
 
 BOOL show_systray = TRUE;
 
-/* an individual systray icon */
-struct tray_icon
-{
-    struct list    entry;
-    HICON          image;    /* the image to render */
-    HWND           owner;    /* the HWND passed in to the Shell_NotifyIcon call */
-    HWND           window;   /* the adaptor window */
-    BOOL           layered;  /* whether we are using a layered window */
-    HWND           tooltip;  /* Icon tooltip */
-    UINT           state;    /* state flags */
-    UINT           id;       /* the unique id given by the app */
-    UINT           callback_message;
-    int            display;  /* display index, or -1 if hidden */
-    WCHAR          tiptext[128];    /* tooltip text */
-    WCHAR          info_text[256];  /* info balloon text */
-    WCHAR          info_title[64];  /* info balloon title */
-    UINT           info_flags;      /* flags for info balloon */
-    UINT           info_timeout;    /* timeout for info balloon */
-    HICON          info_icon;       /* info balloon icon */
-    UINT           version;         /* notify icon api version */
-};
-
 static struct list icon_list = LIST_INIT( icon_list );
 
 static const WCHAR icon_classname[] = {'_','_','w','i','n','e','x','1','1','_','t','r','a','y','_','i','c','o','n',0};
@@ -304,7 +282,7 @@ static void add_to_standalone_tray( struct tray_icon *icon )
 
     icon->display = nb_displayed;
     pos = get_icon_pos( icon );
-    CreateWindowW( icon_classname, NULL, WS_CHILD | WS_VISIBLE,
+    CreateWindowW( icon_classname, icon->tiptext, WS_CHILD | WS_VISIBLE,
                    pos.x, pos.y, icon_cx, icon_cy, standalone_tray, NULL, NULL, icon );
     if (!icon->window)
     {
@@ -668,7 +646,11 @@ static BOOL modify_icon( struct tray_icon *icon, NOTIFYICONDATAW *nid )
     if (nid->uFlags & NIF_TIP)
     {
         lstrcpynW(icon->tiptext, nid->szTip, ARRAY_SIZE( icon->tiptext ));
-        if (icon->tooltip) update_tooltip_text(icon);
+        if (icon->tooltip)
+        {
+            update_tooltip_text(icon);
+            SetWindowTextW(icon->window, icon->tiptext);
+        }
     }
     if (nid->uFlags & NIF_INFO && nid->cbSize >= NOTIFYICONDATAA_V2_SIZE)
     {
diff --git a/dlls/winex11.drv/unixlib.h b/dlls/winex11.drv/unixlib.h
index 7dc1d9f..8226fe6 100644
--- a/dlls/winex11.drv/unixlib.h
+++ b/dlls/winex11.drv/unixlib.h
@@ -17,6 +17,7 @@
  */
 
 #include "ntuser.h"
+#include "wine/list.h"
 #include "wine/unixlib.h"
 
 enum x11drv_funcs
@@ -52,10 +53,32 @@ struct init_params
     BOOL *show_systray;
 };
 
+/* an individual systray icon */
+struct tray_icon
+{
+    struct list    entry;
+    HICON          image;    /* the image to render */
+    HWND           owner;    /* the HWND passed in to the Shell_NotifyIcon call */
+    HWND           window;   /* the adaptor window */
+    BOOL           layered;  /* whether we are using a layered window */
+    HWND           tooltip;  /* Icon tooltip */
+    UINT           state;    /* state flags */
+    UINT           id;       /* the unique id given by the app */
+    UINT           callback_message;
+    int            display;  /* display index, or -1 if hidden */
+    WCHAR          tiptext[128];    /* tooltip text */
+    WCHAR          info_text[256];  /* info balloon text */
+    WCHAR          info_title[64];  /* info balloon title */
+    UINT           info_flags;      /* flags for info balloon */
+    UINT           info_timeout;    /* timeout for info balloon */
+    HICON          info_icon;       /* info balloon icon */
+    UINT           version;         /* notify icon api version */
+};
+
 struct systray_dock_params
 {
     UINT64 event_handle;
-    void *icon;
+    struct tray_icon *icon;
     int cx;
     int cy;
     BOOL *layered;
diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 98b0b23..65e8338 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -2248,7 +2248,8 @@ NTSTATUS x11drv_systray_dock( void *arg )
     *params->layered = layered = (visual.depth == 32);
 
     RtlInitUnicodeString( &class_name, icon_classname );
-    hwnd = NtUserCreateWindowEx( layered ? WS_EX_LAYERED : 0, &class_name, &class_name, NULL,
+    hwnd = NtUserCreateWindowEx( layered ? WS_EX_LAYERED : 0, &class_name, &class_name,
+                                 (UNICODE_STRING*)params->icon->tiptext,
                                  WS_CLIPSIBLINGS | WS_POPUP, CW_USEDEFAULT, CW_USEDEFAULT,
                                  params->cx, params->cy, NULL, 0, NULL, params->icon, 0,
                                  NULL, 0, FALSE );
