diff --git a/src/comgr-compiler.cpp b/src/comgr-compiler.cpp
index 76689cd..acfdc61 100644
--- a/src/comgr-compiler.cpp
+++ b/src/comgr-compiler.cpp
@@ -755,6 +755,8 @@ amd_comgr_status_t AMDGPUCompiler::processFile(const char *InputFilePath,
   if (NoGpuLib)
     Argv.push_back("-nogpulib");
 
+  Argv.push_back("-fno-color-diagnostics");
+
   Argv.push_back("-o");
   Argv.push_back(OutputFilePath);
 
@@ -831,6 +833,7 @@ amd_comgr_status_t AMDGPUCompiler::addIncludeFlags() {
     Args.push_back(PrecompiledHeaderPath.c_str());
     Args.push_back("-Xclang");
     Args.push_back("-fno-validate-pch");
+    Args.push_back("-fno-color-diagnostics");
   }
 
   return AMD_COMGR_STATUS_SUCCESS;
@@ -1107,6 +1110,8 @@ amd_comgr_status_t AMDGPUCompiler::linkToRelocatable() {
     return Status;
   ScopedDataObjectReleaser SDOR(OutputT);
 
+  Args.push_back("-fno-color-diagnostics");
+
   DataObject *Output = DataObject::convert(OutputT);
   Output->setName("a.o");
   auto OutputFilePath = getFilePath(Output, OutputDir);
@@ -1152,6 +1157,8 @@ amd_comgr_status_t AMDGPUCompiler::linkToExecutable() {
     return Status;
   ScopedDataObjectReleaser SDOR(OutputT);
 
+  Args.push_back("-fno-color-diagnostics");
+
   DataObject *Output = DataObject::convert(OutputT);
   Output->setName("a.so");
   auto OutputFilePath = getFilePath(Output, OutputDir);
