diff --git a/runtime/lua/vim/lsp/completion.lua b/runtime/lua/vim/lsp/completion.lua
index 77cd30c87b..9d3b46c7a5 100644
--- a/runtime/lua/vim/lsp/completion.lua
+++ b/runtime/lua/vim/lsp/completion.lua
@@ -303,11 +303,6 @@ function M._lsp_to_complete_items(result, prefix, client_id)
         return match_item_by_value(item.filterText, prefix)
       end
 
-      if item.textEdit then
-        -- server took care of filtering
-        return true
-      end
-
       return match_item_by_value(item.label, prefix)
     end
   end
