diff --git a/src/nvim/mbyte.c b/src/nvim/mbyte.c
index ab787524a..8fd212026 100644
--- a/src/nvim/mbyte.c
+++ b/src/nvim/mbyte.c
@@ -490,12 +490,22 @@ int utf_char2cells(int c)
     if (!utf_printable(c)) {
       return 6;                 // unprintable, displays <xxxx>
     }
+#if 1
+    int n = wcwidth((wchar_t)c);
+    if (n < 0) {
+      return 6;                 // unprintable, displays <xxxx>
+    }
+    if (n > 1) {
+      return n;
+    }
+#else
     if (intable(doublewidth, ARRAY_SIZE(doublewidth), c)) {
       return 2;
     }
     if (p_emoji && intable(emoji_wide, ARRAY_SIZE(emoji_wide), c)) {
       return 2;
     }
+#endif
   } else if (c >= 0x80 && !vim_isprintc(c)) {
     // Characters below 0x100 are influenced by 'isprint' option.
     return 4;                   // unprintable, displays <xx>
