diff --git a/libskk/file-dict.vala b/libskk/file-dict.vala
index 0f5fae0..2db8fa6 100644
--- a/libskk/file-dict.vala
+++ b/libskk/file-dict.vala
@@ -71,8 +71,13 @@ namespace Skk {
         // at the beginning of the next line.
         bool read_until (ref long offset, string line) {
             return_val_if_fail (offset < mmap.length, null);
+            char *p = ((char *)mmap.memory + offset);
+            if (Memory.cmp (p, (void *)line, line.length) == 0) {
+                offset += line.length;
+                return true;
+            }
             while (offset + line.length < mmap.length) {
-                char *p = ((char *)mmap.memory + offset);
+                p = ((char *)mmap.memory + offset);
                 if (*p == '\n' &&
                     Memory.cmp (p + 1, (void *)line, line.length) == 0) {
                     offset += line.length;
