diff --git a/libskk/key-event-filter.vala b/libskk/key-event-filter.vala
index 8c3d6ce..8052da5 100644
--- a/libskk/key-event-filter.vala
+++ b/libskk/key-event-filter.vala
@@ -60,8 +60,6 @@ namespace Skk {
             // ignore key release event
             if ((key.modifiers & ModifierType.RELEASE_MASK) != 0)
                 return null;
-            // clear shift mask
-            key.modifiers &= ~ModifierType.SHIFT_MASK;
             return key;
         }
     }
diff --git a/libskk/key-event.vala b/libskk/key-event.vala
index 195941a..72d4050 100644
--- a/libskk/key-event.vala
+++ b/libskk/key-event.vala
@@ -223,6 +223,9 @@ namespace Skk {
                 if ((modifiers & ModifierType.MOD1_MASK) != 0) {
                     elements.add ("alt");
                 }
+                if ((modifiers & ModifierType.SHIFT_MASK) != 0) {
+                    elements.add ("shift");
+                }
                 if ((modifiers & ModifierType.LSHIFT_MASK) != 0) {
                     elements.add ("lshift");
                 }
