diff --git a/configure.ac b/configure.ac
index c72148d06..dd21e88a4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2609,9 +2609,6 @@ if test x$zsh_cv_c_unicode_support = xyes; then
   AC_CACHE_CHECK(if the wcwidth() and/or iswprint() functions are broken,
   zsh_cv_c_broken_wcwidth,
   [AC_RUN_IFELSE([AC_LANG_SOURCE([[$locale_prog]])],[zsh_cv_c_broken_wcwidth=yes],[zsh_cv_c_broken_wcwidth=no],[zsh_cv_c_broken_wcwidth=no])])
-  if test x$zsh_cv_c_broken_wcwidth = xyes; then
-    AC_DEFINE(ENABLE_UNICODE9)
-  fi
 
   dnl Check if isprint() behaves correctly under UTF-8 locale.
   dnl On some platform (maybe only on Mac OS X), isprint() returns
